api/handler/acl.go

```diff
@@ -1708,11 +1760,87 @@ func isBucketOwnerForced(table *eacl.Table) bool {
 					f.Value() == amzBucketOwnerEnforced &&
 					f.From() == eacl.HeaderFromRequest &&
 					f.Matcher() == eacl.MatchStringNotEqual {
-					return true
+					if len(r.Targets()) == 1 && len(r.Targets()[0].Accounts()) == 1 {
+						return r.Targets()[0].Accounts()[0] == ownerEnforcedUserID
+					}
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func bucketOwnerPreferredRecord() *eacl.Record {
+	var markerRecord = eacl.CreateRecord(eacl.ActionDeny, eacl.OperationPut)
+	markerRecord.AddFilter(
+		eacl.HeaderFromRequest,
+		eacl.MatchStringNotEqual,
+		amzBucketOwnerField,
+		amzBucketOwnerPreferred,
+	)
+
+	t := eacl.NewTarget()
+	t.SetAccounts([]user.ID{ownerPreferredUserID})
+	markerRecord.SetTargets(*t)
+
+	return markerRecord
+}
+
+func isBucketOwnerPreferred(table *eacl.Table) bool {
+	if table == nil {
+		return false
+	}
+
+	for _, r := range table.Records() {
+		if r.Action() == eacl.ActionDeny && r.Operation() == eacl.OperationPut {
+			for _, f := range r.Filters() {
+				if f.Key() == amzBucketOwnerField &&
+					f.Value() == amzBucketOwnerPreferred &&
```

Comment:
Why these filters still if you're using target accounts as well? I thought a single special entity (be it a filter or user) is enough.

Reply:
I had the same thought about it, but internally the gate makes some conversation between EACL <-> AST <-> POLICY formats. Without all these checks it is impossible to save all required context when some ACL is applied to an object/bucket